horovod/common/operations.cc

```diff
@@ -648,28 +665,126 @@ int64_t GetTensorDataForAutotuner(const ResponseList& response_list,
              response.response_type() == Response::ERROR);
 
-      entries.push_back(iter->second);
+      if (!state.joined || state.rank == 0) {
+        // We should never fail at finding this key in the tensor table.
+        auto iter = tensor_table.find(name);
+        assert(iter != tensor_table.end());
 
-      // Clear the tensor table of this tensor and its callbacks; the rest of
-      // this function takes care of it.
-      tensor_table.erase(iter);
+        if (state.joined) {
+          switch (response.tensor_type()) {
+            case HOROVOD_UINT8:
+              iter->second.tensor = std::make_shared<JoinTensor<HOROVOD_UINT8, uint8_t>>(state.join_device, response.tensor_sizes()[0]);
+              break;
+            case HOROVOD_INT8:
+              iter->second.tensor = std::make_shared<JoinTensor<HOROVOD_INT8, int8_t>>(state.join_device, response.tensor_sizes()[0]);
+              break;
+            case HOROVOD_UINT16:
+              iter->second.tensor = std::make_shared<JoinTensor<HOROVOD_UINT16, uint16_t>>(state.join_device, response.tensor_sizes()[0]);
+              break;
+            case HOROVOD_INT16:
+              iter->second.tensor = std::make_shared<JoinTensor<HOROVOD_INT16, int16_t>>(state.join_device, response.tensor_sizes()[0]);
+              break;
+            case HOROVOD_INT32:
+              iter->second.tensor = std::make_shared<JoinTensor<HOROVOD_INT32, int32_t>>(state.join_device, response.tensor_sizes()[0]);
+              break;
+            case HOROVOD_INT64:
+              iter->second.tensor = std::make_shared<JoinTensor<HOROVOD_INT64, int64_t>>(state.join_device, response.tensor_sizes()[0]);
+              break;
+            case HOROVOD_FLOAT16:
+              iter->second.tensor = std::make_shared<JoinTensor<HOROVOD_FLOAT16, float>>(state.join_device, response.tensor_sizes()[0]);
+              break;
+            case HOROVOD_FLOAT32:
+              iter->second.tensor = std::make_shared<JoinTensor<HOROVOD_FLOAT32, float>>(state.join_device, response.tensor_sizes()[0]);
+              break;
+            case HOROVOD_FLOAT64:
+              iter->second.tensor = std::make_shared<JoinTensor<HOROVOD_FLOAT64, double>>(state.join_device, response.tensor_sizes()[0]);
+              break;
+            case HOROVOD_BOOL:
+              iter->second.tensor = std::make_shared<JoinTensor<HOROVOD_BOOL, bool>>(state.join_device, response.tensor_sizes()[0]);
+              break;
+            default:
+              throw std::logic_error("Unknown tensor data type");
+          }
+          iter->second.output = iter->second.tensor;
+          iter->second.device = state.join_device;
+        }
+
+        entries.push_back(iter->second);
+        // Clear the tensor table of this tensor and its callbacks; the rest of
+        // this function takes care of it.
+        tensor_table.erase(iter);
+      } else {
+        TensorTableEntry entry;
+        switch (response.tensor_type()) {
```

Comment:
Extract this switch statement into a helper function, so switch above and here can be deduplicated?

Reply:
Removed the duplication.